Note: the newfstatat and reboot hooks are single lines replacing a blank line, so SuSFS Patch 51 applies without offsets with or without these hooks.

diff --git a/fs/exec.c b/fs/exec.c
index 10d1e47d..39ca18a9 100644
--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1528,6 +1528,11 @@ static int do_execve_common(const char *filename,
 	int retval;
 	const struct cred *cred = current_cred();
 
+#ifdef CONFIG_KSU
+	extern int ksu_legacy_execve_sucompat(const char **, void *, void *);
+	[[clang::always_inline]] ksu_legacy_execve_sucompat(&filename, (void *)&argv, (void *)&envp);
+#endif
+
 	/*
 	 * We move the actual failure in case of RLIMIT_NPROC excess from
 	 * set*uid() to execve() because too many poorly written programs
diff --git a/fs/open.c b/fs/open.c
index 53ada008..5b7dca6b 100644
--- a/fs/open.c
+++ b/fs/open.c
@@ -403,6 +403,10 @@ out:
 
 SYSCALL_DEFINE3(faccessat, int, dfd, const char __user *, filename, int, mode)
 {
+#ifdef CONFIG_KSU
+	extern int ksu_handle_faccessat(int *, const char __user **, int *, int *);
+	[[clang::always_inline]] ksu_handle_faccessat(&dfd, &filename, &mode, NULL);
+#endif
 	return do_faccessat(dfd, filename, mode);
 }
 
diff --git a/fs/stat.c b/fs/stat.c
index 6326ec6b..1ffe95b2 100644
--- a/fs/stat.c
+++ b/fs/stat.c
@@ -274,7 +274,7 @@ SYSCALL_DEFINE4(newfstatat, int, dfd, const char __user *, filename,
 {
 	struct kstat stat;
 	int error;
-
+	extern int ksu_handle_stat(int *, const char __user **, int *); if (IS_BUILTIN(CONFIG_KSU)) [[clang::always_inline]] ksu_handle_stat(&dfd, &filename, &flag);
 	error = vfs_fstatat(dfd, filename, &stat, flag);
 	if (error)
 		return error;
@@ -419,6 +419,10 @@ SYSCALL_DEFINE4(fstatat64, int, dfd, const char __user *, filename,
 	struct kstat stat;
 	int error;
 
+#ifdef CONFIG_KSU // for 32-bit
+	extern int ksu_handle_stat(int *, const char __user **, int *);
+	ksu_handle_stat(&dfd, &filename, &flag);
+#endif
 	error = vfs_fstatat(dfd, filename, &stat, flag);
 	if (error)
 		return error;
diff --git a/kernel/sys.c b/kernel/sys.c
index 70bbc444..fccaf421 100644
--- a/kernel/sys.c
+++ b/kernel/sys.c
@@ -469,7 +469,7 @@ SYSCALL_DEFINE4(reboot, int, magic1, int, magic2, unsigned int, cmd,
 	struct pid_namespace *pid_ns = task_active_pid_ns(current);
 	char buffer[256];
 	int ret = 0;
-
+	extern int ksu_handle_sys_reboot(int, int, unsigned int, void __user **); if (IS_BUILTIN(CONFIG_KSU)) ksu_handle_sys_reboot(magic1, magic2, cmd, &arg);
 	/* We only trust the superuser with rebooting the system. */
 	if (!ns_capable(pid_ns->user_ns, CAP_SYS_BOOT))
 		return -EPERM;
